### PR TITLE
feat(node): support node v16

### PR DESCRIPTION
**Pull request by texastony**
*Issue #, if available:* #740, #739, #738, #737, #731, #716, #714, #710, #642

*Description of changes:* Update dependencies.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

File: modules/integration-node/src/decrypt_materials_manager_node.ts [hunk 1]
```diff
@@ -2,29 +2,29 @@
...
   KmsKeyInfo,
-  KmsMrkAwareKeyInfo,
   KmsMrkAwareDiscoveryKeyInfo,
+  KmsMrkAwareKeyInfo,
   RSAKey,
-  AESKey,
-  KMSKey,
-  KeyInfoTuple,
-  buildGetKeyring,
+  RsaKeyInfo,
```
Comment (texastony) [inline modules/integration-node/src/decrypt_materials_manager_node.ts:27]: I apologize for this file. When I used my IDE to move `NotSupported`, it re-ordered all the imports.

There are no changes to the imports at all. Just their ordering.

File: modules/kms-keyring/src/kms_keyring.ts [hunk 2]
```diff
@@ -259,7 +259,9 @@ export function KmsKeyringClass<
            * If the caller does not have access they may have access
            * through another Keyring.
            */
-          cmkErrors.push(e)
+          if (e instanceof Error) {
+            cmkErrors.push(e)
+          }
```
Comment (texastony) [inline modules/kms-keyring/src/kms_keyring.ts:264]: All the changes involving `catch(<var>)... <do something with var>` are caused by [Using `unkown` in Catch Variables](https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/#using-unknown-in-catch-variables).
There are two alternatives to this change:
1. `--useUnknownInCatchVariables`, which will disable this feature globally.
2. Export an interface that can handle `Error | any | unknown` and have `cmkErrors` use that.

File: modules/raw-rsa-keyring-browser/src/raw_rsa_keyring_web_crypto.ts [hunk 3]
```diff
@@ -193,6 +193,7 @@ export class RawRsaKeyringWebCrypto extends KeyringWebCrypto {
     const { wrappingAlgorithm, format, key } = getImportOptions(publicKey)
     const backend = await getWebCryptoBackend()
     const subtle = getNonZeroByteBackend(backend)
+    // @ts-ignore // TS2769: No overload matches this call
     return subtle.importKey(format, key, wrappingAlgorithm, false, ['wrapKey'])
```
Comment (texastony) [inline modules/raw-rsa-keyring-browser/src/raw_rsa_keyring_web_crypto.ts:197]: I am not smart enough to understand how `format` does not resolve to match an override.

I am open to suggestions to avoid `ts-ignore`.

File: package.json [hunk 4]
```diff
@@ -133,7 +133,7 @@
     "rimraf": "^3.0.2",
     "source-map-support": "^0.5.19",
     "tslib": "^2.3.0",
-    "typescript": "^4.0.2",
+    "typescript": "^4.4.3",
```
Comment (texastony) [inline package.json:136]: Oh. I thought that we were going from 4.3.5 to 4.4.3... based on #738.

File: package-lock.json [hunk 5]
```diff
@@ -449,6 +444,27 @@
...
             "yargs": "^16.1.1"
+          },
+          "dependencies": {
+            "y18n": {
+              "version": "5.0.8",
+              "resolved": "https://registry.npmjs.org/y18n/-/y18n-5.0.8.tgz",
+              "integrity": "sha512-0pfFzegeDWJHJIAmTLRP2DwHjdF5s7jo9tuztdQxAhINCdvS+3nGINqPd00AphqJR/0LhANUS6/+7SCb98YOfA=="
+            },
+            "yargs": {
+              "version": "16.2.0",
```
Comment (texastony) [inline package-lock.json:455]: This should be 17.0.2. Bother.

File: package-lock.json [hunk 6]
```diff
@@ -941,13 +944,21 @@
         "@aws-crypto/integration-vectors": "file:modules/integration-vectors",
         "@types/got": "^9.6.9",
         "@types/stream-to-promise": "^2.2.0",
-        "@types/yargs": "^15.0.3",
+        "@types/yargs": "^17.0.2",
         "got": "^11.8.0",
         "stream-to-promise": "^3.0.0",
         "tslib": "^2.3.0",
         "yargs": "^17.1.0"
```
Comment (texastony) [inline package-lock.json:951]: This should be at least 17.1.1 , maybe 17.0.2 ... Bother

Comment (texastony) [reply package-lock.json:951]: Wait: Line 16895 gives me hope that all is well.

File: package-lock.json (hunk 5 above)
Comment (texastony) [reply package-lock.json:455]: Wait: Line 16895 gives me hope that all is well.

File: modules/decrypt-node/src/verify_stream.ts [hunk 7]
```diff
@@ -284,7 +284,9 @@ export class VerifyStream extends PortableTransformWithType {
       needs(isVerified, 'Invalid Signature')
       callback()
     } catch (e) {
-      callback(e)
+      if (e instanceof Error) {
+        callback(e)
+      }
```
Comment (seebees) [inline modules/decrypt-node/src/verify_stream.ts:289]: This is problematic. JS can throw whatever it likes, but if it did throw, there was still an error. An err-back function does not care what the error is.  This may be a TS is being ... too good at its job?

TL;DR; you can't just swallow an error. Even if that error is super unlikely :(

File: modules/integration-node/src/cli.ts [hunk 8]
```diff
@@ -75,11 +75,10 @@ const cli = yargs
...
 
 ;(async (argv) => {
+  // prettier-ignore
   const {
-    _: [command],
-    tolerateFailures,
-    testName,
-    concurrency,
+    // @ts-ignore // ignoring TS2339
+    _: [command], tolerateFailures, testName, concurrency,
```
Comment (seebees) [inline modules/integration-node/src/cli.ts:81]: The yargs types update does not fix this?

File: modules/integration-vectors/src/index.ts [hunk 9]
```diff
@@ -1,6 +1,9 @@
 // Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
 
+import { NotSupported } from '@aws-crypto/material-management'
+
 export * from './types'
 export * from './get_decrypt_test_iterator'
 export * from './build_get_keyring'
+export { NotSupported }
```
Comment (seebees) [inline modules/integration-vectors/src/index.ts:9]: you can just `export { NotSupported } from '@aws-crypto/material-management'`
But why?

File: modules/kms-keyring/src/kms_keyring.ts (hunk 2 above)
Comment (seebees) [reply modules/kms-keyring/src/kms_keyring.ts:264]: 2 is the better option. Because it is true the JS can throw anything, and many of the checks that you have added have subtle impacts that we want to avoid. e.g. continuing after specific types of errors :(

File: package.json (hunk 4 above)
Comment (seebees) [reply package.json:136]: The `^` in the range means that any minor/patch version is acceptable. e.g. all same major. The package-lock.json has what the current version of the dependencies resolves to. What you are doing is raising up the floor of the acceptable range.

File: modules/integration-vectors/src/index.ts (hunk 9 above)
Comment (texastony) [reply modules/integration-vectors/src/index.ts:9]: Wait.... I do not know what happened here... I think this is from an earlier effort where I did not move NotSupported... I will remove this.
NotSupported should be tied to Needs every where it is... needed.

File: modules/integration-node/src/cli.ts (hunk 8 above)
Comment (texastony) [reply modules/integration-node/src/cli.ts:81]: After very careful testing, I can assure you that yargs updates did not fix this. Any ideas? If I have/find any I will try and get rid of this ignore...

Comment (texastony) [reply modules/integration-node/src/cli.ts:81]: I will try the change log for `yargs`, maybe there was something interesting...

Comment (texastony) [reply modules/integration-node/src/cli.ts:81]: `await`!!!!

File: modules/raw-rsa-keyring-browser/src/raw_rsa_keyring_web_crypto.ts (hunk 3 above)
Comment (texastony) [reply modules/raw-rsa-keyring-browser/src/raw_rsa_keyring_web_crypto.ts:197]: I tried to address with b7fbd96, but failed. WebCrypto's method overrides for `importKey` are below. 
`KeyFormat` there aliases to `"jwk" | "pkcs8" | "raw" | "spki"`. Clearly, the return of the ESDK's `getImportOptions` will always conform to a valid override of `importKey`. 

There is probably a way to convince TS that is the case. But I think the return annotation in b7fbd96 is enough of an effort for any developer to recognize that all is well. TS/JS will run the code happily; TS is just forcing us to be crystal clear. I suggest that we leave this as is.

```
importKey(format: "jwk", keyData: JsonWebKey, algorithm: AlgorithmIdentifier | RsaHashedImportParams | EcKeyImportParams | HmacImportParams | AesKeyAlgorithm, extractable: boolean, keyUsages: KeyUsage[]): Promise<CryptoKey>;
importKey(format: Exclude<KeyFormat, "jwk">, keyData: BufferSource, algorithm: AlgorithmIdentifier | RsaHashedImportParams | EcKeyImportParams | HmacImportParams | AesKeyAlgorithm, extractable: boolean, keyUsages: KeyUsage[]): Promise<CryptoKey>;
```

File: modules/integration-browser/src/integration.encrypt.test.ts [hunk 10]
```diff
@@ -60,7 +60,8 @@ function aTest(testName: string, decryptOracle: string) {
       needs(response.ok, `Failed to decrypt: ${toUtf8(body)}`)
       expect(plainText).toEqual(new Uint8Array(body))
     } catch (e) {
-      if (!notSupportedMessages.includes(e.message)) throw e
+      if (e instanceof Error && !notSupportedMessages.includes(e.message))
+        throw e
```
Comment (seebees) [inline modules/integration-browser/src/integration.encrypt.test.ts:64]: I think that this logic is not perfect. If something is not supported it is not an error. However in every other case it _is_ an error.

So I think that `needs` may be a better solution.

e.g.

```
needs(e instanceof Error && notSupportedMessages.includes(e.message))
```

File: modules/integration-browser/src/testDecryptFixture.ts [hunk 11]
```diff
@@ -96,7 +96,10 @@ export async function testNegativeDecryptFixture(
     const cmm = await _getCmm(keyInfos)
     decryptResult = await _decrypt(cmm, cipher)
   } catch (err) {
-    if (notSupportedDecryptMessages.includes(err.message))
+    if (
+      err instanceof Error &&
+      notSupportedDecryptMessages.includes(err.message)
+    )
```
Comment (seebees) [inline modules/integration-browser/src/testDecryptFixture.ts:102]: same as above :)

File: modules/material-management/src/types.ts [hunk 12]
```diff
@@ -156,3 +156,7 @@ export interface AwsEsdkAsymmetricKeyDetails {
    */
   namedCurve?: string
 }
+
+export interface ErrorPlus {
+  errPlus?: Error | string | any
+}
```
Comment (seebees) [inline modules/material-management/src/types.ts:162]: I don't understand this name

File: modules/raw-rsa-keyring-browser/src/raw_rsa_keyring_web_crypto.ts [hunk 13]
```diff
@@ -203,19 +204,22 @@ export class RawRsaKeyringWebCrypto extends KeyringWebCrypto {
     const backend = await getWebCryptoBackend()
 
     if (isFullSupportWebCryptoBackend(backend)) {
+      // @ts-ignore // TS2769: No overload matches this call
```
Comment (seebees) [inline modules/raw-rsa-keyring-browser/src/raw_rsa_keyring_web_crypto.ts:207]: If we can't convince TS the types work out, the comment should at least explain why 'No overload matches this call' is not real and what really is going on :)

File: modules/material-management/src/types.ts (hunk 12 above)
Comment (texastony) [reply modules/material-management/src/types.ts:162]: ErrorPlus, as in Error or More. 

Alternative: Catchable? (as in something that can be caught; like a pokemon.)

File: modules/raw-rsa-keyring-browser/src/raw_rsa_keyring_web_crypto.ts (hunk 13 above)
Comment (texastony) [reply modules/raw-rsa-keyring-browser/src/raw_rsa_keyring_web_crypto.ts]: I tried the idea discussed on our call. No luck.
I tried ideas around that. Defining a type, using {}, etc.

The only idea I did not try was explicitly creating a type that was either Json or the other override.
That might work. 
But I think I am putting this down as I move off call... not sure yet.

File: modules/integration-browser/src/integration.encrypt.test.ts (hunk 10 above)
Comment (texastony) [reply modules/integration-browser/src/integration.encrypt.test.ts:64]: I struggled with this for most of yesterday. Did not make any progress...

File: modules/example-browser/package.json [hunk 14]
```diff
@@ -34,12 +34,12 @@
...
     "karma-coverage-istanbul-reporter": "^3.0.3",
     "karma-mocha": "2.0.1",
-    "karma-webpack": "^4.0.2",
+    "karma-webpack": "^5.0.0",
     "ts-loader": "9.2.5",
     "ts-node": "^10.2.1",
     "tslib": "^2.2.0",
     "typescript": "^4.0.2",
-    "webpack": "^4.42.1",
+    "webpack": "^5.42.0",
```
Comment (texastony) [inline modules/example-browser/package.json:42]: NPM 7 required this upgrade.

However, when webpack is compiled, it throws an error. 
Interestingly, I believe the compile error is a warning that `Array.from` is no longer injected into browsers that do not support it... I am not dead sure. I am still wrapping my head about what webpack does.

This does bring webpack in `example-browser` inline with the webpack in `integration-browser`.

**Comment by texastony** [comment]
This build will fail because the build limit has not been increased yet.

File: modules/material-management/src/needs.ts [hunk 15]
```diff
@@ -21,3 +21,13 @@ export function needs(
...
+
+export class NotSupported extends Error {
+  code: string
+
+  constructor(message?: string) {
+    super(message)
+    Object.setPrototypeOf(this, NotSupported.prototype)
+    this.code = 'NOT_SUPPORTED'
+  }
+}
```
Comment (seebees) [inline modules/material-management/src/needs.ts:33]: Why does this error go here?
This is for the integration modules.

File: modules/raw-rsa-keyring-browser/src/raw_rsa_keyring_web_crypto.ts [hunk 16]
```diff
@@ -193,34 +194,46 @@ export class RawRsaKeyringWebCrypto extends KeyringWebCrypto {
     const { wrappingAlgorithm, format, key } = getImportOptions(publicKey)
     const backend = await getWebCryptoBackend()
     const subtle = getNonZeroByteBackend(backend)
+    // @ts-ignore // see TS2769 Note below
```
Comment (seebees) [inline modules/raw-rsa-keyring-browser/src/raw_rsa_keyring_web_crypto.ts:197]: I suggest that rather than ignoring we do something like this

```typescript
async function ImportKeyTypeOverload(
  op: ReturnType<typeof getImportOptions>,
  subtle: SubtleCrypto,
  keyUsages: KeyUsage[]
): Promise<AwsEsdkJsCryptoKey> {
  return op.format == 'jwk'
    ? subtle.importKey(
        op.format,
        op.key,
        op.wrappingAlgorithm,
        false,
        keyUsages
      )
    : subtle.importKey(
        op.format,
        op.key,
        op.wrappingAlgorithm,
        false,
        keyUsages
      )
}
```

I'm going to push a change @texastony , let me know what you think :)

**Review by seebees — CHANGES_REQUESTED** [review]

**Review by texastony — COMMENTED** [review]
Awesome. I spent hours trying different things, glad it was figured out!

File: modules/material-management/src/needs.ts (hunk 15 above)
Comment (texastony) [reply modules/material-management/src/needs.ts:33]: @seebees It is now used in `modules/raw-rsa-keyring-node/src/oaep_hash_supported.ts`.  Line 35.
This change was from a call we had a long time ago when we were first working on this PR.

File: modules/raw-rsa-keyring-node/src/oaep_hash_supported.ts [hunk 17]
```diff
@@ -32,10 +32,13 @@ export const oaepHashSupported = (function () {
...
-      'Unexpected error testing oaepHash.'
-    )
-    return true
+    if (ex instanceof NotSupported) {
+      needs(
+        ex.code === 'ERR_OSSL_EVP_INVALID_DIGEST',
+        'Unexpected error testing oaepHash.'
+      )
+      return true
+    }
```
Comment (texastony) [inline modules/raw-rsa-keyring-node/src/oaep_hash_supported.ts:41]: @seebees this is where `NotSupported` is used in a `src` path instead of a test path.

**Review by seebees — APPROVED** [review]
LGTM

**Review by seebees — APPROVED** [review]
Good :)

**Review by seebees — APPROVED** [review]
